### PR TITLE
Adds working Helpscout > App > SFDC connection

### PR DESCRIPTION
**Pull request by Salisbum**
@jpurcell001  and @lee-feigenbaum for review and initial comments before I continue with snagging and formatting SFDC content.

These commits are the base of the rest of the application.
- Establishes that requests are verified as originating from Help Scout
- Creates new SFDC object that initializes a Restforce client
- Returns a formatted `html.erb` file containing the SFDC object's `customer_email` to ensure proper content display

### CODE REVIEW

File: app/models/helpscout.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
+require 'json'
+
+class HelpScout
+  def initialize()
```
Comment (jpurcell001) [inline app/models/helpscout.rb]: Ruby - no parentheses, and also since this is an empty body, no need for the method at all.

File: app/models/helpscout.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  end
+
+  def parse_message_for_user(body)
+    content = JSON.parse(body)
+
+    if content['customer'] && content['customer']['email']
```
Comment (jpurcell001) [inline app/models/helpscout.rb]: In general I'm a big fan of using Hashie::Mash for situations like this.  Combined with the lonely operator in Ruby 2.3, it lets you do this as
```
content = Hashie::Mash.new(JSON.parse(body))
content&.customer&.email
```

File: Gemfile [hunk 3]
```diff
@@ -0,0 +1,14 @@
+source 'https://rubygems.org'
+
+gem 'pry'
+gem 'rake'
+gem 'sinatra'
+gem 'sinatra-activerecord'
+gem 'restforce', '~> 2.4.1'
+
+ruby '2.3.0'
```
Comment (jpurcell001) [inline Gemfile]: Go with latest version when possible - 2.3.1

And best to put this at or near top of file to make it clear, rather than being buried in between gems

File: Gemfile [hunk 4]
```diff
@@ -0,0 +1,14 @@
+source 'https://rubygems.org'
+
+gem 'pry'
```
Comment (jpurcell001) [inline Gemfile]: Do you need `pry` outside of development?

File: app/models/helpscout.rb [hunk 5]
```diff
@@ -0,0 +1,16 @@
+require 'json'
+
+class HelpScout
```
Comment (jpurcell001) [inline app/models/helpscout.rb]: Some comments on the code for reference but in reality the comments above about the class design really trump all this.

File: app/models/helpscout.rb [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+  def initialize()
+  end
+
+  def parse_message_for_user(body)
+    content = JSON.parse(body)
+
+    if content['customer'] && content['customer']['email']
+      customer_email = content['customer']['email']
+    end
+    customer_email
```
Comment (jpurcell001) [inline app/models/helpscout.rb]: This isn't necessary.  The whole `customer_email` isn't necessary.  Even if you didn't do it as the one-liner suggested above, Ruby methods return the last value - if the if block didn't fire then it would be nil, otherwise it would be the value inside the if block.

File: app/models/salesforce_record.rb [hunk 7]
```diff
@@ -0,0 +1,10 @@
+require 'restforce'
+
+class SalesforceRecord
```
Comment (jpurcell001) [inline app/models/salesforce_record.rb]: Really not clear what this class is meant to do.  Obviously it's a stub at this point but the naming, etc. just is unclear.  I'm guessing you want to have a class where you can initialize with an email and get at details about a contact, org, account, etc.  If that's the case I would have something more like

```
class Person

def initialize(email)
  @email = email
end

def organization
 # ...
end

def account
# ...
end

# ...

def salesforce
  @salesforce ||= Restforce::Client.new
end
```

(Naming will be easier once you have a better handle on how you'll use this)

Remember that Restforce returns salesforce objects as instances of Hashie::Mash so it's really easy to work with them e.g. `account.Name`

File: app/views/sfdc_content.html.erb [hunk 8]
```diff
@@ -0,0 +1,10 @@
+<!DOCTYPE html>
+<html>
+  <head>
```
Comment (jpurcell001) [inline app/views/sfdc_content.html.erb]: Is any of this necessary?

File: app.rb [hunk 9]
```diff
@@ -0,0 +1,40 @@
+require 'sinatra'
+require_relative 'config/application'
+require 'json'
+require 'base64'
+require 'openssl'
+
+post '/helpscout.salesforce' do
```
Comment (jpurcell001) [inline app.rb]: I would see if you can rethink the OO design of this.

It's really not clear what instances of class HelpScout are, what instances of class SalesforceRecord are, etc.  

It sort of feels like what you want is a class called 'HelpscoutConversation' or 'HelpScout::Conversation'.  Each instance of this class would represent a message from Helpscout (per the doc [here](http://developer.helpscout.net/custom-apps/dynamic/), it sounds like the payload represents a conversation).  
You could initialize instances of this class by passing in a body and parsing there e.g.
`conversation = HelpscoutConversation.new(request.body.read)`
```
def initialize(body)
  @conversation_data = Hashie::Mash.new(JSON.parse(body))
end
```
You could also have a class message that takes a request and verifies signature:
`HelpscoutConversation.verify_signature!(request)`
This would allow you to move all the signature verification logic into a Helpscout specific class.  
Instances of a helpscout conversation could expose useful methods for getting at the data e.g. the customer email:
```
def customer_email
   @conversation_data&.customer&.email
end
```

Your verify_signature! method (! suggests it throws an exception if it fails) can then be moved into a before filter on the sinatra controller, to slim down the request handling even further.  

I'd take a crack at cleaning up the OO a bit - make sure that each instance of a class represents something easy to describe.  And think about how you would go about testing each independently (even if you don't have time to write the tests now - just reasoning through how you WOULD write the tests is usually helpful)

File: app.rb [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+  verify_signature
+
+  body = request.body.read
+  helpscout = HelpScout.new
+
+  if customer_email = helpscout.parse_message_for_user(body)
+    @sfdc_record = SalesforceRecord.new(customer_email)
```
Comment (jpurcell001) [inline app.rb]: Feels like you're jumping through a lot of hoops here to do something that Sinatra would give you by default.  Can't you just use their templates natively, combined with a helper method to get a handle on the HelpscoutConversation object?  (If that doesn't make sense, take a look at http://www.sinatrarb.com/intro.html#Views%20/%20Templates)

File: app/models/salesforce_record.rb (hunk 7 above)
Comment (jpurcell001) [inline app/models/salesforce_record.rb]: I might even just have these be methods of HelpscoutConversation - not sure that this warrants a whole another class.

File: app.rb (hunk 9 above)
Comment (Salisbum) [inline app.rb]: @jpurcell001 I've made the above requested changes with some exceptions. 

Re: the `verify_signature` method (and correct me if I'm misunderstanding)- Moving it into the `HelpscoutConversation` class makes sense as we are verifying the request incoming from that specific convo. 

However, in order to do something like 
```
before do
   verify_signature!
end
```

I would need to move that method back out of the `HelpscoutConversation` and into `app.rb` - I don't feel comfortable running it on a newly initialized `HelpscoutConversation` instance as the content hasn't been verified yet (under the assumption `#initialize` would contain the `body`, which I have not done in this commit).

Tl;DR - I would rather verify the source before saving the `request.body` to an instance of `HelpscoutConversation`.

Alternatively, this is an option:
```
before do
  @helpscout = HelpscoutConversation.new

  @helpscout.verify_signature
end

post '/helpscout.salesforce' do
  body = request.body.read
  #some stuff here
end
```

Comment (jpurcell001) [inline app.rb]: Re: the views:  wouldn't it be easier to just use the idiomatic way to render sinatra templates?

E.g.
```
if conversation.email
  erb :sfdc_content, locals: { conversation: conversation }
else
  'Can't find this user.'
end
```

Might need to play with the specifics to get it right (I haven't used Sinatra templates before) but check out http://www.sinatrarb.com/intro.html#Return%20Values and http://www.sinatrarb.com/intro.html#Views%20/%20Templates.

File: app/models/helpscout_conversation.rb [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+    bail and return false unless (sig = env['HTTP_X_HELPSCOUT_SIGNATURE']).present?
+
+    digest = OpenSSL::Digest.new('sha1')
+
+    hmac = OpenSSL::HMAC.new(ENV.fetch('SECRET_KEY'), digest)
+
+    bail and return false unless sig.strip == Base64.encode64(hmac.digest).strip
+  end
+
+  def bail
```
Comment (jpurcell001) [inline app/models/helpscout_conversation.rb]: Also isn't this whole method just boil down to a Sinatra oneliner
`halt 403, 'Bad Signature.'`
?

File: app.rb (hunk 9 above)
Comment (Salisbum) [inline app.rb]: I did another pass through regarding these comments - there's some fun erroring when trying to implement a `!` for `verify_signature`.

Additionally, I had to move both the `verify_signature` and `bail` methods back to the controller as the response was not working properly from within the `HelpscoutConversation` class. 

Regarding the simplification of the views, Sinatra doesn't seem to have a more elegant way of rendering `erd` views as json in a response outside the means I'm currently using that I was able to find via the resources you provided and Google.

I'm happy to go through my logic for this structure with you in person, but I'd like to get this project completed ASAP for Ethan and also before I go on vacation to manage my apartment move later this week.
